backend/src/database/postgres/PostgresDatabase.ts

```diff
@@ -0,0 +1,111 @@
+import { Pool, PoolClient, types } from 'pg';
+import { BaseCategory, Category } from '../../models/category';
+import { BaseExpense, Expense } from '../../models/expense';
+import { BaseIncome, Income } from '../../models/income';
+import { BaseUser, User } from '../../models/user';
+import { Database } from '../database';
+
+export class PgDatabase implements Database {
+  private static singleton: PgDatabase | null;
+  private static readonly pool = new Pool();
+
+  static instance(): PgDatabase {
+    if (this.singleton == null) {
+      this.singleton = new PgDatabase();
+    }
+    return this.singleton;
+  }
+
+  private constructor() {
+    // For when query results parse postgres types to typescript types
+    types.setTypeParser(23, parseInt); // 23 represents int types in pg
+    types.setTypeParser(701, parseFloat); // 701 represents double precision in pg
+  }
+
+  private async transaction<T>(
+    action: (c: PoolClient, commit: () => Promise<void>, rollback: () => Promise<void>) => Promise<T>,
+  ): Promise<T> {
+    const client = await PgDatabase.pool.connect();
+    await client.query('BEGIN');
+    const transactionResult = action(
+      client,
+      async () => {
+        await client.query('COMMIT');
+      },
+      async () => {
+        await client.query('ROLLBACK');
+      },
+    );
+    client.release();
+    return transactionResult;
+  }
```

Comment:
We can use this transaction method to open and close a transaction with the action callback (where we actually do the queries)